### PR TITLE
Fixes #4060

### LINKED ISSUE #4060: Gzip::Writer causing broken pipe
I just updated Kemal for Crystal 0.21.0 and one of the breaking changes is the `Gzip::Writer`, `Flate::Writer`

As reported by @crisward (https://github.com/kemalcr/kemal/issues/316) `Gzip::Writer` causes broken pipe on `0.21.0` but used to work as `Zlib::Deflate.gzip` before.

Here are the line https://github.com/kemalcr/kemal/blob/master/src/kemal/static_file_handler.cr#L64-L66 

The stacktrace

``` 
2017-02-22 17:13:10 +0300 500 GET /bundle.js 30.35ms
Unhandled exception in spawn:
Error writing file: Broken pipe (Errno)
0x103fd2582: *CallStack::unwind:Array(Pointer(Void)) at ??
0x103fd2521: *CallStack#initialize:Array(Pointer(Void)) at ??
0x103fd24f8: *CallStack::new:CallStack at ??
0x103fc4891: *raise<Errno>:NoReturn at ??
0x1040f0ce4: *TCPSocket+@IO::FileDescriptor#unbuffered_write<Slice(UInt8)>:Int32 at ??
0x1040f09a1: *TCPSocket+@IO::Buffered#flush:TCPSocket+ at ??
0x1040f107f: *TCPSocket+@IO::Buffered#close:Nil at ??
0x1040db3a9: *HTTP::Server::RequestProcessor#process<(OpenSSL::SSL::Socket::Server | TCPSocket+), (OpenSSL::SSL::Socket::Server | TCPSocket+), IO::FileDescriptor>:Nil at ??
0x1040d28a5: *HTTP::Server::RequestProcessor#process<(OpenSSL::SSL::Socket::Server | TCPSocket+), (OpenSSL::SSL::Socket::Server | TCPSocket+)>:Nil at ??
0x1040cd366: *HTTP::Server#handle_client<(TCPSocket+ | Nil)>:Nil at ??
0x103fcff36: ~procProc(Nil)@/usr/local/Cellar/crystal-lang/0.21.0/src/http/server.cr:163 at ??
0x103fdcc24: *Fiber#run:(IO::FileDescriptor | Nil) at ??
0x103fca3e9: ~proc2Proc(Fiber, (IO::FileDescriptor | Nil))@/usr/local/Cellar/crystal-lang/0.21.0/src/fiber.cr:29 at ??
```

### PR DESCRIPTION
**Pull request by crisward**
Extra byte evaluated to truthy with an empty slice. Added size check which fixes the gzip bug.
I'll remove my previous pull request.

### CODE REVIEW

File: src/gzip/header.cr [hunk 1]
```diff
@@ -71,7 +73,7 @@ class Gzip::Header
 
     # flg
     flg = Flg::None
-    flg |= Flg::EXTRA if @extra
+    flg |= Flg::EXTRA if !@extra.empty?
```
Comment (RX14) [inline src/gzip/header.cr]: Use `unless`

File: src/gzip/header.cr [hunk 2]
```diff
@@ -85,8 +87,8 @@ class Gzip::Header
     # os
     io.write_byte os
 
-    if extra = @extra
-      io.write_byte extra.size.to_u8
+    if !@extra.empty?
```
Comment (RX14) [inline src/gzip/header.cr]: Ditto

File: src/gzip/header.cr (hunk 1 above)
Comment (crisward) [inline src/gzip/header.cr]: Have done...

Comment (crisward) [inline src/gzip/header.cr]: thanks

File: src/gzip/header.cr (hunk 2 above)
Comment (crisward) [inline src/gzip/header.cr]: Ditto

**Comment by ysbaddaden** [comment]
Thanks, I was about to suggest the same (or `if @extra.any?` to be positive)

**Comment by asterite** [comment]
No, this fix is a workforce, the real issue is a compiler bug. @extra is always initialized, it has an initializer

File: src/gzip/header.cr [hunk 3]
```diff
@@ -44,6 +44,8 @@ class Gzip::Header
       xlen = io.read_byte.not_nil!
       @extra = Bytes.new(xlen)
       io.read_fully(@extra)
+    else
+      @extra = Bytes.empty
```
Comment (asterite) [inline src/gzip/header.cr]: This line is not needed, `@extra` should always be initialized. Right now it doesn't work because of #3988

**Comment by asterite** [comment]
Sorry, I closed it too soon. The fix is OK, the but adding an `else; @extra = Bytes.empty` should be removed from this PR.

**Comment by crisward** [comment]
Ok will update.

**Comment by RX14** [comment]
Oh, actually :+1: to `if @extra.any?`.

**Comment by asterite** [comment]
`Enumerable#any?` determines if any of the elements is truthy. That's definitely slower and more complex than checking for `empty?`. Please change it :-)

**Comment by crisward** [comment]
Changed it back to empty - this only runs once per gzip operation... doubt the speed difference would be significant compared to everything else in the zipping process, but I was happy with `if !@extra.empty?` so what do I know?

**Comment by asterite** [comment]
@crisward Ah, my comment is misleading indeed, sorry. The point is that `any?` means "is any element in this byte slice truthy?" which is not exactly what you want here. Here we just need to know if the byte slice is empty. Using `any?` makes the semantic unclear for a reader that is not familiar with the code.

**Comment by crisward** [comment]
Ok, new to ruby/crystal so didn't know what `any` meant, assumed with was the opposite to empty. Originally had it as `.size > 0`. Also many other language don't have unless.. Ah well, hopefully all this will be squashed into a nice merge assuming the orange turns to green.

**Comment by RX14** [comment]
Oh, sorry @asterite, I just assumed that `any?` was `!empty?` because of @ysbaddaden's comment.

**Comment by ysbaddaden** [comment]
Oh, I always assumed `any?` was just the negation of `empty?`. I was wrong.

**Comment by asterite** [comment]
In practice it works the same if the elements will always be truthy. The problem is that `any?` is very short so it's very tempting to use it instead of `!empty?` 😄

File: src/gzip/header.cr [hunk 4]
```diff
@@ -85,8 +85,8 @@ class Gzip::Header
     # os
     io.write_byte os
 
-    if extra = @extra
-      io.write_byte extra.size.to_u8
+    unless @extra.empty?
+      io.write_byte @extra.size.to_u8
       io.write(extra)
```
Comment (Sija) [inline src/gzip/header.cr]: shouldn't this also refer to `@extra` for consistency?

Comment (crisward) [inline src/gzip/header.cr]: Think you're right. It used to copy the value, I'm guessing their isn't a test case for gzip with extra set. I'll update.

**Comment by ukd1** [comment]
This sounds like it will also fix #4064

**Comment by spalladino** [comment]
Related to #4084 as well?

**Pull request merged**